gdal/frmts/jpegxr/jpegxrdataset.cpp

```diff
@@ -0,0 +1,1202 @@
+        if (OVERLAP_NONE < 0 || n > OVERLAP_COUNT)
+        {
+            CPLError(CE_Failure, CPLE_IllegalArg,
+                "OVERLAP=%s is not a legal value in the range %d-%d.",
+                CSLFetchNameValue(papszOptions, "OVERLAP"), OVERLAP_NONE, OVERLAP_COUNT-1);
+            return nullptr;
+        }
+        config.SetOverlap(n);
+    }
+
+    if (CSLFetchNameValue(papszOptions, "SUBSAMPLING") != NULL)
+    {
+        int n = atoi(CSLFetchNameValue(papszOptions, "SUBSAMPLING"));
+        if (SUBSAMPLING_420 < 0 || n > SUBSAMPLING_COUNT)
+        {
+            CPLError(CE_Failure, CPLE_IllegalArg,
+                "SUBSAMPLING=%s is not a legal value in the range %d-%d.",
+                CSLFetchNameValue(papszOptions, "SUBSAMPLING"), SUBSAMPLING_420, SUBSAMPLING_COUNT-1);
+            return nullptr;
+        }
+        config.SetSubsampling(n);
+    }
+    /* -------------------------------------------------------------------- */
+    /*      Read source data                                                */
+    /* -------------------------------------------------------------------- */
+    int nWordSize = GDALGetDataTypeSize(eDT) / 8;
+    int nUncompressedSize = nXSize * nYSize * nBands * nWordSize;
+    GByte* pabyUncompressedData = (GByte*)VSI_MALLOC_VERBOSE(nUncompressedSize);
+    if (!pabyUncompressedData)
+    {
+        VSIFree(pabyUncompressedData);
+        return nullptr;
+    }
+
+    CPLErr eErr;
+    eErr = poSrcDS->RasterIO(GF_Read, 0, 0, nXSize, nYSize,
+        pabyUncompressedData, nXSize, nYSize,
+        eDT, nBands, NULL,
+        nBands * nWordSize, nBands * nWordSize * nXSize, nWordSize, NULL);
+    if (eErr != CE_None)
+    {
+        VSIFree(pabyUncompressedData);
+        return nullptr;
+    }
+
+    /* -------------------------------------------------------------------- */
+    /*      Setup encoder                                                   */
+    /* -------------------------------------------------------------------- */
+    try
+    {
+        JPEGXREncoder oEncoder;
+        oEncoder.Initialize(pszFilename);
+        oEncoder.SetVerbose(false); // TODO: option?
+        oEncoder.SetSize(nXSize, nYSize);
+        oEncoder.SetResolution(72, 72);
+
+        if (nBands == 1 && poSrcDS->GetRasterBand(1)->GetColorInterpretation() == GCI_GrayIndex)
+        {
+            oEncoder.SetPixelFormat(GUID_PKPixelFormat8bppGray);
+        }
+        else if (nBands >= 3)
+        {
+            // RGB
+            if (poSrcDS->GetRasterBand(1)->GetColorInterpretation() == GCI_RedBand &&
+                poSrcDS->GetRasterBand(2)->GetColorInterpretation() == GCI_GreenBand &&
+                poSrcDS->GetRasterBand(3)->GetColorInterpretation() == GCI_BlueBand)
+            {
+                if (nBands == 3)
+                    oEncoder.SetPixelFormat(GUID_PKPixelFormat24bppRGB);
+                else
+                    oEncoder.SetPixelFormat(GUID_PKPixelFormat32bppRGBA);
```

Comment:
In theory there might be rasters with 4 bands that are RGBsomething_else. Might be useful to check against GCI_AlphaBand. Same for BGRA